libs2eplugins/src/s2e/Plugins/Models/BaseFunctionModels.cpp

```diff
@@ -81,16 +82,21 @@ bool BaseFunctionModels::findNullCharWithWidth(S2EExecutionState *state, uint64_
 
         bool truth;
         bool res = solver->mustBeTrue(query, truth);
-        if (res && truth) {
+
+        if (truth&&res) {
+            allSymbolic = false; // Found a null byte, not all characters are symbolic
+            getDebugStream(state) << "Found nullptr at offset " << len << "\n";
             break;
         }
     }
 
-    if (len == MAX_STRLEN) {
-        getDebugStream(state) << "Could not find nullptr char\n";
-        return false;
+    if (allSymbolic && len == MAX_STRLEN) {
+        // If we reached the end and all characters were symbolic, null-terminate the string
+        uint8_t *str = reinterpret_cast<uint8_t *>(stringAddr);
+        str[len - width] = '\0'; 
```

Comment:
This should not work. The address is from the guest, you cannot directly dereference. Please add a test case to the testsuite to verify what you write is correct.

Reply:
Yes, this way segfaults s2e. I changed to use executionStateMemory's write to update the guest memory